File: 611.ValidTriangleNumber/Solution.py
class Solution:
    def triangleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if nums is None or len(nums)<3:
            return 0
        
        #sort list
        nums.sort()
        count = 0
        for third_edge_idx in range(2,len(nums)):
            #third_edge = 2, 3, ... n-1
            first_edge_idx, second_edge_idx = 0, third_edge_idx-1
            while first_edge_idx < second_edge_idx:
                if nums[first_edge_idx] + nums[second_edge_idx] > nums[third_edge_idx]:
                    count += second_edge_idx - first_edge_idx
                    second_edge_idx -= 1
                else:
                    first_edge_idx += 1
                    
        return count
                

File: 611.ValidTriangleNumber/Solution.cpp
class Solution {
  public:
    int triangleNumber(vector < int > & nums) {

      if (nums.size() < 3) return 0;
      sort(nums.begin(), nums.end());
      int count = 0;
      for (int third_edge = 2; third_edge < nums.size(); third_edge++) {
        int first_edge = 0, second_edge = third_edge - 1;
        while (first_edge < second_edge) {
          if (nums[first_edge] + nums[second_edge] > nums[third_edge]) {
            count += second_edge - first_edge;
            second_edge--;
          } else {
            first_edge++;
          }
        }
      }
      return count;
    }
};